### DISCUSSION TITLE
Is it possible to flatten nested models in a type-safe way?

### DISCUSSION THREAD

**Discussion by CobaltCause** [pydantic/pydantic#4236] [Question] [answered]
Does `pydantic` have an analog to [`#[serde(flatten)]`](https://serde.rs/field-attrs.html#flatten)? For example, I'd like the following behavior:

```python
class Foo(BaseModel):
    foo: str


class Bar(BaseModel):
    bar: str


class Qux(BaseModel):
    foo: Foo # = Field(flatten=True) # or something
    bar: Bar # = Field(flatten=True) # or something


foo = Foo(foo="blah")
bar = Bar(bar="blah blah")
qux = Qux(foo=foo, bar=bar)

expected = {
    "foo": "blah",
    "bar": "blah blah",
}

assert qux.dict() == expected
```

Specifically, I want to avoid doing this:

```python
class Foo(BaseModel):
    foo: str


class Bar(BaseModel):
    bar: str


class Qux(Foo, Bar):
    pass


foo = Foo(foo="blah")
bar = Bar(bar="blah blah")
qux = Qux(**foo.dict(), **bar.dict())  # no thanks

expected = {
    "foo": "blah",
    "bar": "blah blah",
}

assert qux.dict() == expected
```

because this cannot be statically typechecked to ensure that `foo` and `bar` are actually `Foo` and `Bar`.

**Comment by PrettyWood** [comment, marked as answer]
Hi @CobaltCause!
No there is no such thing currently unfortunately. But it's interesting and may well be added in v2 (cc @samuelcolvin)
In the meantime here is a workaround

```py
from pydantic import BaseModel, Field


class MyBaseModel(BaseModel):
    def _iter(self, to_dict: bool = False, *args, **kwargs):
        for dict_key, v in super()._iter(to_dict, *args, **kwargs):
            if to_dict and self.__fields__[dict_key].field_info.extra.get('flatten', False):
                assert isinstance(v, dict)
                yield from v.items()
            else:
                yield dict_key, v

class Foo(MyBaseModel):
    foo: str


class Bar(MyBaseModel):
    bar: str


class Qux(MyBaseModel):
    foo: Foo = Field(flatten=True)
    bar: Bar = Field(flatten=True)


foo = Foo(foo="blah")
bar = Bar(bar="blah blah")
qux = Qux(foo=foo, bar=bar)

assert qux.dict() == {
    "foo": "blah",
    "bar": "blah blah",
}
assert qux.foo.foo == "blah"
assert qux.bar.bar == "blah blah"
```

Hope it helps!

**Reply by CobaltCause** [reply]
Cool trick! Though I suppose I should have mentioned that I'd like deserialization and `Qux.schema_json` to work as well in the same way. Either way, thanks!

**Reply by PeterNociar** [reply]
Thank you!! Using this I was able to make something like marshmallow's fields.Pluck to get a single value from a nested model:
```python
user_name: User = Field(pluck='name')
```


``` python
def _iter(self, to_dict: bool = False, *args, **kwargs):
  for dict_key, v in super()._iter(to_dict, *args, **kwargs):
    if to_dict:
      if pluck_field := self.__fields__[dict_key].field_info.extra.get('pluck', False):
        yield dict_key, v.get(pluck_field)
    yield dict_key, v

**Comment by samuelcolvin** [comment]
https://pydantic-docs.helpmanual.io/blog/pydantic-v2/#more-powerful-aliases explains how it'll work on V2

**Reply by CobaltCause** [reply]
Exciting! Thanks for the great library btw!

**Comment by marcowartmann** [comment]
Hi all

Is there already an easy way to flatten a nested model while model_dump_json in v2? And if yes, could you please show an example?

**Comment by danish-cashify** [comment]
is there anyway to make it openapi friendly?

**Reply by sebpretzer** [reply]
[datamodel-code-generator](https://github.com/koxudaxi/datamodel-code-generator/) is the way that [pydantic recommends](https://docs.pydantic.dev/latest/integrations/datamodel_code_generator/#code-generation-with-datamodel-code-generator) you to do the conversion
